### ISSUE TITLE
Consider adding some of the nontrivial traversal things

### ISSUE THREAD

**Issue by neongreen** [monadfix/microlens#44]
Should add: `taking`, `dropping`, `element`, `elementOf`.

Done with: `failing`, `filtered`, `singular`.

**Comment by neongreen** [comment]
Added `failing`.

**Comment by neongreen** [comment]
Added `filtered`.

**Comment by neongreen** [comment]
Added `singular`.

**Comment by fosskers** [comment]
As long as `microlens` stays `micro` ;)

**Comment by neongreen** [comment]
@fosskers: I checked, it still builds in 4s – and that's despite me having to copy an implementation of `StateT` into the library.

**Comment by fosskers** [comment]
How long before this was added? Still 4s?

**Comment by neongreen** [comment]
No, about 3.6s.

Once it goes past 4s, I guess I'd have to do something like splitting `Lens.Micro` into 2 modules internally (so that GHC would be able to compile them in parallel), but I'm not worried about it now.

**Comment by libscott** [comment]
If you split it into two packages you could call one of them "Lens.NonMicro" and make it optional :).

**Comment by neongreen** [comment]
Hm, darn. I checked and it looks like splitting Lens.Micro into several modules doesn't help much (it wins us 0.1s).

**Comment by fosskers** [comment]
Where are we at now for build times?

**Comment by neongreen** [comment]
I started using `stack build` instead of `cabal build` to measure build times (mainly because it makes it easier to measure total build times for packages that have dependencies – you can see the table of build times for all microlens packages in the readme), and for some reason it's faster with Stack. Specifically, building microlens takes 3s (and it's 2.9s with split modules and `--ghc-options -j4`).

**Comment by neongreen** [comment]
I'm not going to add anything else without a user request.
